presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -335,7 +335,18 @@ public PlanNode visitUnion(UnionNode node, RewriteContext<Expression> context)
         @Override
         public PlanNode visitFilter(FilterNode node, RewriteContext<Expression> context)
         {
-            return context.rewrite(node.getSource(), combineConjuncts(node.getPredicate(), context.get()));
+            PlanNode rewrittenPlan = context.rewrite(node.getSource(), combineConjuncts(node.getPredicate(), context.get()));
+            if (!(rewrittenPlan instanceof FilterNode)) {
```

Comment:
what else can it be here? maybe use `verify` instead?

Reply:
filter could be consumed by table scan or pushed below aggregation, join